core/src/main/java/com/linecorp/armeria/common/stream/FilteredStreamMessage.java

```diff
@@ -190,10 +191,14 @@ public void onSubscribe(Subscription s) {
         @Override
         public void onNext(T o) {
             ReferenceCountUtil.touch(o);
-            if (!withPooledObjects) {
+            if (!filterSupportsPooledObjects) {
```

Comment:
I'm confusing, is this parameter really necessary for `FilteredStreamMessage`? As you said [this comment](https://github.com/line/armeria/pull/1770#issuecomment-493003539), this is not a final subscriber, so it seems unnecessary to care whether the buffer is pooled or not from here. Let me think about this more tomorrow. 🤔

Reply:
For our current implementations, nope but if a user implemented their own filtering response, it may not be aware of pooled objects, so the input to `filter` would be leaked. I think that's why we have the default that the filter does not support pooled objects, and an advanced user would have to opt-in with this parameter.

Reply:
Yeah, I agree with you on that. :)

Reply:
Basically I also agree with @anuraaga's comment. 👍 Then, don't we need to handle the case which might copy the buffer twice?

I'm not sure, but is it okay for here?
```
if (!filterSupportsPooledObjects || !subscribedWithPooledObjects) {
    o = PooledObjects.toUnpooled(o);
}
```
, or like this?
```
if (!subscribedWithPooledObjects &&
    (o != filtered || filterSupportsPooledObjects)) {
    filtered = PooledObjects.toUnpooled(filtered);
}
```
What do you think? @minwoox @anuraaga

Reply:
Ah that's a good point! We shouldn't call toUnpooled if we already did.
So it could be:
```java
if (!filterSupportsPooledObjects) {
    o = PooledObjects.toUnpooled(o);
}
U filtered = filter(o);
if (filterSupportsPooledObjects && !subscribedWithPooledObjects) {
    filtered = PooledObjects.toUnpooled(filtered);
}
```
?

Reply:
Calling `toUnpooled` on an unpooled object unnecessarily does a couple of extra instance checks but seems safe. I thought about that conditional but don't think there is any hard contract that just because an implementation didn't pass `withPooledObjects`, meaning it's input is never pooled, it wasn't allowed to publish pooled objects. If we wanted to make this our behavior, we should update the javadoc of the parameter to explicitly state this (it currently only talks about input). The overhead for unpooling an unpooled one isn't so bad so we could keep it simpler or add that javadoc - what do you think?

Reply:
Ok. Let's leave it as it is. Thank you!

Reply:
That makes sense. Thank!